### PR TITLE
Command Handler

### PR DESCRIPTION
**Pull request by LamboCreeper**
Creates a command handler to register commands for the bot. Any commands that have the property `visible` set to `true` will also appear in a command list.

An example command should have code like (you can use this as a template for building commands if you wish):
```js
// Set the command's properties
const properties = {
	command: "test",
	description: "This describes what the command does.",
	prefix: true,
	arguments: ["<required argument>", "[optional argument]"],
	visible: false,
	botchat: false
};

// The code that runs when the command is executed.
function run(message) {
	console.log(`${message.author.username} ran the command.`);
}

// Export the data to the command handler
exports.properties = properties
exports.run = run;
```

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -2,15 +2,76 @@
...
+const fs = require("fs");
 const Discord = require("discord.js");
 
 // Setup the Discord Client
 const client = new Discord.Client({
 	fetchAllMembers: true
 });
 
+// Create an object to store the command in.
+var commands = {};
```
Comment (Bas-Sarrechia) [inline app.js]: Use let.

File: app.js [hunk 2]
```diff
@@ -2,15 +2,76 @@
 const config = require("./config.json");
 
 // Package Dependencies
+const fs = require("fs");
```
Comment (Bas-Sarrechia) [inline app.js]: what is 'fs'. Const are written in capital with underscores

File: app.js [hunk 3]
```diff
@@ -2,15 +2,76 @@
 const config = require("./config.json");
```
Comment (itsiffy) [inline app.js]: Const are written in capital with underscores

File: app.js [hunk 4]
```diff
@@ -2,15 +2,76 @@
...
 
 // Package Dependencies
+const fs = require("fs");
 const Discord = require("discord.js");
 
 // Setup the Discord Client
 const client = new Discord.Client({
```
Comment (itsiffy) [inline app.js]: Const are written in capital with underscores

File: app.js [hunk 5]
```diff
@@ -2,15 +2,76 @@
...
+		}
+	});
+});
+
+// Run when a message is sent
+client.on("message", (message) => {
+	for (const commandFile in commands) {
+		const command = commands[commandFile];
+		const properties = command.properties;
+		const args = message.content.split(" ");
```
Comment (itsiffy) [inline app.js]: Const are written in capital with underscores

File: app.js [hunk 6]
```diff
@@ -2,15 +2,76 @@
...
+					} else {
+						command.run(message, args);
+					}
+				}	
+			}
+		}
+	}
+
+	if (message.content.startsWith(`${config.prefix}commands`)) {
+		const embed = new Discord.RichEmbed();
```
Comment (itsiffy) [inline app.js]: Const are written in capital with underscores

File: app.js [hunk 7]
```diff
@@ -2,15 +2,76 @@
...
+
+		embed.setTitle("Commands");
+		for (const commandFile in commands) {
+			const properties = commands[commandFile].properties;
```
Comment (itsiffy) [inline app.js]: Const are written in capital with underscores

File: app.js [hunk 8]
```diff
@@ -2,15 +2,76 @@
...
+
+			if (properties.visible) {
+				const command = properties.prefix ? `${config.prefix}${properties.command}` : properties.command;
+				const arguments = properties.arguments.length == 0 ? "" : properties.arguments.toString().replace(new RegExp(",", "g"), " ");
+
```
Comment (itsiffy) [inline app.js]: Const are written in capital with underscores

**Review by Bas-Sarrechia — CHANGES_REQUESTED** [review]
see inline comments.

File: app.js (hunk 2 above)
Comment (LamboCreeper) [reply app.js]: It's Node.js's file I/O system, it is used for reading the commands directory as part of the command handler.

Comment (Bas-Sarrechia) [reply app.js]: Require a meaningful non abriviated but short name for this variable.

**Review by Bas-Sarrechia — APPROVED** [review]

**Comment by saramaebee** [comment]
I still disagree with the usage of `fileSystem` instead of `fs`. I think this should be discussed more with everyone who would be working on the bot, and I won't approve it until a consensus is reached.

File: app.js [hunk 9]
```diff
@@ -2,15 +2,75 @@
...
+					}
+				}	
+			} else {
+				if (message.content.startsWith(properties.command)) {
+					if (properties.arguments.length == 0) {
+						command.run(message);
+					} else {
+						command.run(message, args);
+					}
+				}
```
Comment (Jason2605) [inline app.js:49]: Could this section not be shortened down to:
```js
const messagePrefix = properties.prefix ? config.prefix : properties.command;

if (message.content.startsWith(messagePrefix)) {
    if (properties.arguments.length === 0) {
        command.run(message);
    } else {
        command.run(message, args);
    }
}
```

**Review by saramaebee — APPROVED** [review]
Approving. Discussions regarding `fs` vs `fileSystem` needs to occur. If it is decided that we do not use `fs`, `fileSystem` should be changed to `filesystem`

**Pull request merged**